File: scripts/create_dictionary.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(data.table)
  library(stringr)
  library(parallel)
  library(dplyr)
  library(purrr)
  library(tidyr)
})

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 5) {
  stop("Usage: Rscript create_dictionary.R [collapsed_file] [haplotag_dir] [classification_file] [threads] [output_dir]")
}

collapsed_file       <- args[1]
haplotag_dir         <- args[2]
classification_file  <- args[3]
n_threads            <- as.integer(args[4])
output_dir           <- args[5]

dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

collapsed_data <- fread(collapsed_file, header = TRUE, col.names = c("read_id", "isoform"))
classification_data <- fread(classification_file, header = TRUE, sep = "\t",
                             select = c("isoform", "structural_category", "associated_gene", "associated_transcript", "subcategory"))

hap <- list.files(haplotag_dir, pattern = "haplotagged.txt", full.names = TRUE) %>%
  set_names() %>%  
  map_dfr(~ fread(.x, header = TRUE, stringsAsFactors = FALSE, check.names = TRUE) %>%
            mutate(Filename = .x), .id = "Filename")

isoform_merge <- left_join(collapsed_data, classification_data) 
isoform_merge <- left_join(isoform_merge, hap, by = join_by("read_id" == "ReadName"))

dictionary <- isoform_merge %>%
    mutate(
    sample = str_extract(read_id, "PS\\d+"),
    condition = str_extract(read_id, "treated|untreated")
  )

counts_hap <- dictionary %>%
  group_by(sample, isoform, condition, Haplotype) %>%
  count() %>%
  mutate(n_hap_cond = n) %>%
  pivot_wider(id_cols = c(sample, isoform, Haplotype), names_from = c(condition), values_from = n_hap_cond) %>%
  mutate(iso_hap_noncyclo_counts = paste0("hn:i:", `non-cyclo`),
         iso_hap_cyclo_counts = paste0("hc:i:", cyclo)) %>%
  select(-cyclo, -`non-cyclo`)

counts <- dictionary %>%
  group_by(sample, isoform, condition) %>%
  count() %>%
  mutate(n_cond = n) %>%
  pivot_wider(c(sample, isoform), names_from = condition, values_from = n_cond) %>%
  mutate(iso_noncyclo_counts = paste0("sn:i:", `non-cyclo`),
         iso_cyclo_counts = paste0("sc:i:", cyclo)) %>%
  select(-cyclo, -`non-cyclo`)

dictionary <- left_join(dictionary, counts_hap) %>%
  left_join(counts) %>%
  mutate(across(everything(), ~replace(., is.na(.), 0))) %>%
  mutate(isoform = gsub("PB.", "in:Z:", isoform),
         structural_category = paste0("sc:Z:", structural_category),
         associated_gene = paste0("gn:Z:", associated_gene),
         associated_transcript = paste0("tn:Z:", associated_transcript),
         subcategory = paste0("sb:Z:", subcategory)) %>%
  distinct()

samples_list <- unique(dictionary$sample)

mclapply(samples_list, function(sample_id) {
  dictionary_out <- dictionary %>%
    filter(sample == sample_id) %>%
    select(-ReadQuality, -ReadLength, -Haplotype, -Filename, -condition, -id)
  
  output_file <- file.path(output_dir, paste0("dictionary_", sample_id, ".txt"))
  write.table(dictionary_out, output_file, quote = FALSE, col.names = TRUE, row.names = FALSE, sep = "\t")
}, mc.cores = n_threads)

File: scripts/create_dictionary.py
#!/usr/bin/env python3

import sys
import os
import re
import glob
import pandas as pd

# ------------------
# 1) Helper functions
# ------------------

def process_collapsed(filepath):
    """
    Replicates R's process_collapsed():
      read.table(filepath, header=TRUE, col.names=c("read_id","isoform"))
    """
    # If the file truly has headers, you can use header=True and remove the names=...
    # Adjust if your input file is formatted differently.
    collapsed_data = pd.read_csv(filepath, sep="\t", header=0,
                                 names=["read_id", "isoform"])
    return collapsed_data


def process_classification(filepath):
    """
    Replicates R's process_classification():
      read.table(..., header=TRUE, sep="\t") %>%
      select(isoform, structural_category, associated_gene, associated_transcript, subcategory)
    """
    classification_data = pd.read_csv(filepath, sep="\t", header=1)
    classification_data = classification_data[
        ["isoform", "structural_category",
         "associated_gene", "associated_transcript", "subcategory"]
    ]
    return classification_data


def create_dictionary_sample(collapsed_data, classification_data, hap_data):
    """
    Replicates R's create_dictionary_sample() steps:
      1) semi_join collapsed/hap_data
      2) semi_join classification/collapsed_sub
      3) left_join everything
      4) fill NA with 0
      5) rename_with(tolower)
      6) string replacements
      7) distinct()
      8) final select(...)
      9) pivot/wider for condition stats
      10) final merges with pivoted stats
    """

    # 1) collapsed_sub <- collapsed_data %>%
    #      semi_join(hap_data, by=join_by("read_id"=="ReadName"))
    # In Python, replicate semi_join via .isin():
    collapsed_sub = collapsed_data[
        collapsed_data["read_id"].isin(hap_data["ReadName"])
    ]

    # 2) classification_sub <- classification_data %>%
    #      semi_join(collapsed_sub, by="isoform")
    classification_sub = classification_data[
        classification_data["isoform"].isin(collapsed_sub["isoform"])
    ]

    # 3) dictionary <- collapsed_sub %>%
    #      left_join(classification_sub, by="isoform") %>%
    #      left_join(hap_data, by=join_by("read_id"=="ReadName"))
    dictionary = pd.merge(
        collapsed_sub,
        classification_sub,
        on="isoform",
        how="left"
    )
    dictionary = pd.merge(
        dictionary,
        hap_data,
        left_on="read_id",
        right_on="ReadName",
        how="left"
    )

    # 4) fill NA with 0 (replicate mutate_all(~ replace(., is.na(.), 0)))
    dictionary = dictionary.fillna(0)

    # 5) rename_with(tolower)
    dictionary.columns = [col.lower() for col in dictionary.columns]

    # 6) string replacements
    dictionary["isoform"] = dictionary["isoform"].astype(str).str.replace("PB.", "in:Z:")
    dictionary["structural_category"] = "sc:Z:" + dictionary["structural_category"].astype(str)
    dictionary["associated_gene"] = "gn:Z:" + dictionary["associated_gene"].astype(str)
    dictionary["associated_transcript"] = "tn:Z:" + dictionary["associated_transcript"].astype(str)
    dictionary["subcategory"] = "sb:Z:" + dictionary["subcategory"].astype(str)

    # 7) distinct()
    dictionary = dictionary.drop_duplicates()

    # 8) final select(...)
    # NOTE: If your haplotag data does NOT have "haplotype" or "condition" or "sample",
    # you must create them or remove them here.
    dictionary = dictionary[
        ["read_id", "haplotype", "isoform", "condition", "sample",
         "structural_category", "associated_gene", "associated_transcript", "subcategory"]
    ]

    # 9) Build additional stats (pivot/wider as in R)
    #    counts_hap -> pivot -> iso_hap_noncyclo_counts, iso_hap_cyclo_counts
    counts_hap = (
        dictionary.groupby(["sample", "isoform", "condition", "haplotype"])
        .size()
        .reset_index(name="n_hap_cond")
    )

    counts_hap_wide = counts_hap.pivot_table(
        index=["sample", "isoform", "haplotype"],
        columns="condition",
        values="n_hap_cond",
        fill_value=0
    ).reset_index()

    if "cyclo" not in counts_hap_wide.columns:
        counts_hap_wide["cyclo"] = 0
    if "non-cyclo" not in counts_hap_wide.columns:
        counts_hap_wide["non-cyclo"] = 0

    counts_hap_wide["iso_hap_noncyclo_counts"] = "hn:i:" + counts_hap_wide["non-cyclo"].astype(str)
    counts_hap_wide["iso_hap_cyclo_counts"] = "hc:i:" + counts_hap_wide["cyclo"].astype(str)
    counts_hap_wide = counts_hap_wide.drop(columns=["cyclo", "non-cyclo"])

    # counts -> pivot -> iso_noncyclo_counts, iso_cyclo_counts
    counts = (
        dictionary.groupby(["sample", "isoform", "condition"])
        .size()
        .reset_index(name="n_cond")
    )

    counts_wide = counts.pivot_table(
        index=["sample", "isoform"],
        columns="condition",
        values="n_cond",
        fill_value=0
    ).reset_index()

    if "cyclo" not in counts_wide.columns:
        counts_wide["cyclo"] = 0
    if "non-cyclo" not in counts_wide.columns:
        counts_wide["non-cyclo"] = 0

    counts_wide["iso_noncyclo_counts"] = "hn:i:" + counts_wide["non-cyclo"].astype(str)
    counts_wide["iso_cyclo_counts"] = "hc:i:" + counts_wide["cyclo"].astype(str)
    counts_wide = counts_wide.drop(columns=["cyclo", "non-cyclo"])

    # 10) final merges with pivoted stats
    dictionary = pd.merge(
        dictionary,
        counts_hap_wide,
        on=["sample", "isoform", "haplotype"],
        how="left"
    )
    dictionary = pd.merge(
        dictionary,
        counts_wide,
        on=["sample", "isoform"],
        how="left"
    )

    return dictionary


# -------------------------
# 2) Main script logic
# -------------------------

def main():
    # Parse CLI arguments
    args = sys.argv[1:]
    collapsed_file = args[0]
    haplotag_dir = args[1]
    classification_file = args[2]

    print("collapsed_file:", collapsed_file)
    print("haplotag_dir:  ", haplotag_dir)
    print("classification:", classification_file)

    # Load main data
    collapsed_data = process_collapsed(collapsed_file)
    classification_data = process_classification(classification_file)

    # Create output folder if needed
    os.makedirs("tag", exist_ok=True)

    # Locate all haplotagged files
    haplotag_files = glob.glob(os.path.join(haplotag_dir, "*haplotagged.txt"))
    print(f"Found {len(haplotag_files)} haplotag files.")

    # Build small DataFrame of file_path, sample
    df_files = pd.DataFrame({"file_path": haplotag_files})

    # replicate R's mutate(sample = sub("^(.*)_(?:treated|untreated).*", "\\1"))
    df_files["sample"] = df_files["file_path"].apply(
        lambda x: re.sub(r"^(.*)_(?:treated|untreated).*", r"\1", os.path.basename(x))
    )
    df_files["sample"] = df_files["sample"].str.replace(r"whatshap//", "", regex=True)

    # group by sample
    files_by_sample = df_files.groupby("sample")["file_path"].apply(list).reset_index()

    for _, row in files_by_sample.iterrows():
        sample_name = row["sample"]
        these_files = row["file_path"]
        print("Processing sample:", sample_name)
        print("  Haplotag files for this sample:", these_files)

        hap_data_list = []
        for f in these_files:
            # Read data
            tmp = pd.read_csv(f, sep="\t", header=0)

            # Check if empty
            if tmp.empty:
                print(f"File {f} has no data (empty). Skipping.")
                continue

            # replicate R mutate calls
            tmp["sample"] = tmp["ReadName"].str.replace(
                r"^(.*?)_(?:treated|untreated)_.*", r"\1", regex=True
            )
            tmp["condition"] = tmp["ReadName"].apply(
                lambda rn: "cyclo" if "_cyclo" in rn
                else ("non-cyclo" if "_non-cyclo" in rn else None)
            )
            tmp["identifier"] = tmp["ReadName"].str.extract(r"(PS[^_]+)")

            hap_data_list.append(tmp)

        # If all files are empty => skip creating dictionary
        if not hap_data_list:
            print(f"No data for sample {sample_name}. Skipping dictionary creation.")
            continue

        hap_data = pd.concat(hap_data_list, ignore_index=True)

        # Create dictionary for this sample
        dict_one = create_dictionary_sample(collapsed_data, classification_data, hap_data)

        # Write out
        out_name = f"tag/dictionary_{sample_name}.txt"
        dict_one.to_csv(out_name, sep="\t", index=False)
        print("Wrote:", out_name)

    print("Done!")


if __name__ == "__main__":
    main()
